pkg/common/cns-lib/volume/manager.go

```diff
@@ -565,63 +576,184 @@ func (m *defaultManager) DeleteVolume(ctx context.Context, volumeID string, dele
 			log.Errorf("ConnectCns failed with err: %+v", err)
 			return err
 		}
+		if m.idempotencyHandlingEnabled {
+			return m.deleteVolumeWithIdempotency(ctx, volumeID, deleteDisk)
+		}
+		return m.deleteVolumeWithoutIdempotency(ctx, volumeID, deleteDisk)
+	}
+	start := time.Now()
+	err := internalDeleteVolume()
+	if err != nil {
+		prometheus.CnsControlOpsHistVec.WithLabelValues(prometheus.PrometheusCnsDeleteVolumeOpType,
+			prometheus.PrometheusFailStatus).Observe(time.Since(start).Seconds())
+	} else {
+		prometheus.CnsControlOpsHistVec.WithLabelValues(prometheus.PrometheusCnsDeleteVolumeOpType,
+			prometheus.PrometheusPassStatus).Observe(time.Since(start).Seconds())
+	}
+	return err
+}
+
+// deleteVolumeWithoutIdempotency attempts to delete the volume on CNS. CNS task information is not persisted and callers
+// cannot assume that operations are idempotent.
+func (m *defaultManager) deleteVolumeWithoutIdempotency(ctx context.Context, volumeID string, deleteDisk bool) error {
+	log := logger.GetLogger(ctx)
+	// Construct the CNS VolumeId list.
+	var cnsVolumeIDList []cnstypes.CnsVolumeId
+	cnsVolumeID := cnstypes.CnsVolumeId{
+		Id: volumeID,
+	}
+	// Call the CNS DeleteVolume.
+	cnsVolumeIDList = append(cnsVolumeIDList, cnsVolumeID)
+	task, err := m.virtualCenter.CnsClient.DeleteVolume(ctx, cnsVolumeIDList, deleteDisk)
+	if err != nil {
+		if cnsvsphere.IsNotFoundError(err) {
+			log.Infof("VolumeID: %q, not found, thus returning success", volumeID)
+			return nil
+		}
+		log.Errorf("CNS DeleteVolume failed from the  vCenter %q with err: %v", m.virtualCenter.Config.Host, err)
+		return err
+	}
+	// Get the taskInfo.
+	taskInfo, err := cns.GetTaskInfo(ctx, task)
+	if err != nil || taskInfo == nil {
+		log.Errorf("failed to get DeleteVolume taskInfo from vCenter %q with err: %v",
+			m.virtualCenter.Config.Host, err)
+		return err
+	}
+	log.Infof("DeleteVolume: volumeID: %q, opId: %q", volumeID, taskInfo.ActivationId)
+	// Get the task results for the given task.
+	taskResult, err := getTaskResultFromTaskInfo(ctx, taskInfo)
+	if err != nil {
+		log.Errorf("failed to get DeleteVolume task result with error: %v", err)
+		return err
+	}
+	volumeOperationRes := taskResult.GetCnsVolumeOperationResult()
+	if volumeOperationRes.Fault != nil {
```

Comment:
I think its possible that we may get NotFound fault from the task result as well. Any reason why we are not checking it?

Reply:
I've reached out to CNS about this. Since this is the existing behavior i wont change it for now

Reply:
So. how can we track it so that we do not forget? Does it make sense to file a GitHub issue for now?